### PR TITLE
Feature/issue 344 asof join memory efficiency

### PR DESCRIPTION
**Pull request by chenziliang**
PR checklist:
- Did you run ClangFormat ?
- Did you separate headers to a different section in existing community code base ?
- Did you surround `proton: starts/ends` for new code in existing community code base ?

Please write user-readable short description of the changes:

### CODE REVIEW

**Comment by chenziliang** [comment]
Micro-benchmark shows 10x memory reduction for single row per chunk case.

File: src/Core/Settings.h [hunk 1]
```diff
@@ -792,7 +792,8 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(Bool, enable_backfill_from_historical_store, true, "Enable backfill data from historical data store", 0) \
     M(Bool, emit_aggregated_during_backfill, true, "Enable emit intermediate aggr result during backfill historical data", 0) \
     M(Bool, include_internal_streams, false, "Show internal streams on SHOW streams query.", 0) \
-    M(UInt64, join_max_cached_bytes, 524288000, "Max cached bytes for stream to stream join", 0) \
+    M(UInt64, join_max_buffered_bytes, 524288000, "Max buffered bytes for stream to stream join", 0) \
+    M(UInt64, join_buffered_data_block_size, 0, "For streaming join, when buffered data in memory, the data block size directs to merge small data blocks to form bigger ones to improve memory efficiency. 0 means disable merging small data blocks.", 0) \
```
Comment (yl-lisen) [inline src/Core/Settings.h:796]: I prefer setting name like this `join_max_rows_for_compact_buffered_data`

File: src/Interpreters/Streaming/benchmarks/data_blocks_perf.cpp [hunk 2]
```diff
@@ -296,111 +296,134 @@ BENCHMARK_CAPTURE(
...
+    /*data_block_size=*/0)
+    ->Iterations(1);
+
+BENCHMARK_CAPTURE(
+    refCountDataBlockList,
+    Column3ChunkRows1Chunks10000000BS1000,
+    /*chunk_columns=*/3,
+    /*chunk_rows=*/1,
+    /*chunks*=*/10'000'000,
+    /*data_block_size=*/1000)
```
Comment (yl-lisen) [inline src/Interpreters/Streaming/benchmarks/data_blocks_perf.cpp:345]: Here we only test one `data_block_size = 1000`
For rows = 1, it seems that the larger the `data_block_size`, the smaller the memory cost.

File: src/Core/Settings.h (hunk 1 above)
Comment (chenziliang) [reply src/Core/Settings.h:796]: We follow the 'max_block_size' convention here.

File: src/Interpreters/Streaming/benchmarks/data_blocks_perf.cpp (hunk 2 above)
Comment (chenziliang) [reply src/Interpreters/Streaming/benchmarks/data_blocks_perf.cpp:345]: In general, this is true. And we can emulate other combinations in this test and there are lots of combinations. 1000 is just a good starting point for this case. And this is another reason the data block size setting is exposed to end users for now.

**Pull request merged**